# Exploring eBay Used Car Sales Data

### Introduction

We will analyze a dataset of used car sales from the classifieds section of the German eBay website. The original dataset was scraped and uploaded to Kaggle by the user orgesleka.
Dataquest(a website that provides courses on data analysis) did some modifications to the original dataset for learning purposes.
The source of the dataset that we are currently working on is [here](https://data.world/data-society/used-cars-data)

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In [66]:
#importing pandas

import pandas as pd

In [67]:
#reading the data file

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [68]:
autos.info()

autos.head()

#checking no. of columns and rows
autos.shape 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

(50000, 20)

We notice the following discrepancies:
- Some columns have missing values (vehicleType, model, fuelType etc)
- Some columns have wrong dtype (price, odometer)
- Column name follow CamelCase style instead of snake_style

** Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive. **

In [69]:
#printing columns

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [70]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [71]:
#checking column names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [72]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

**Observations:**
- The price and odometer columns should be a numerical data type
- registration_year column has unrealistic min and max values!!! (min = 1000 ,max = 9999).
- registration_month column has unrealistic min value (min = 0).
- There are multiple text columns with almost all of the values are the same (seller, offer_type, gear_box, etc)

In [73]:
#price column before conversion
autos['price'].head(10)

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
5    $7,900
6      $300
7    $1,990
8      $250
9      $590
Name: price, dtype: object

In [74]:
#to convert this to int we need to remove the '$' and ',' sign and then change the dtype

autos['price'] = autos['price'].str.strip('$').str.replace(',', '').astype(float)


In [75]:
#checking the price columns
autos['price'].head(10)

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
5    7900.0
6     300.0
7    1990.0
8     250.0
9     590.0
Name: price, dtype: float64

In [76]:
#odometer column before conversion
autos['odometer'].head(10)

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
5    150,000km
6    150,000km
7    150,000km
8    150,000km
9    150,000km
Name: odometer, dtype: object

In [77]:
autos['odometer'] = autos['odometer'].str.strip('km').str.replace(',','').astype(int)

In [78]:
#checking the odometer columns after conversion
autos['odometer'].head(10)

0    150000
1    150000
2     70000
3     70000
4    150000
5    150000
6    150000
7    150000
8    150000
9    150000
Name: odometer, dtype: int64

In [79]:
#renaming odometer column name to odometer_km

autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns.

In [80]:
#Checking unique values in odometer_km
autos['odometer_km'].unique().shape



(13,)

In [81]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [82]:
autos['odometer_km'].value_counts().sort_index(ascending=False)


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

No outlier spotted in the Odometer columns. 
Moving to price columns

In [83]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [85]:
autos['price'].value_counts().sort_index(ascending=False).head(10)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
Name: price, dtype: int64

In real life, no one is ready to pay millions of dollars to buy a used car through the internet!
We will remove the outlier values that are over or equal to 1$ million.
So,the removal range is between 999,999 - 99,999,999

In [86]:
outlier_price = autos[autos['price'].between(999999.0,99999999.0)]

In [88]:
#number of outlier prices (13)
outlier_price.shape

(13, 20)

In [89]:
autos.drop(outlier_price.index, inplace = True)

In [91]:
#now checking the minimum value
autos['price'].value_counts().sort_index(ascending=True).head(10)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
Name: price, dtype: int64

About 1,421 prices as 0, no price in life is equal to 0 we will remove them.
While the prices that are equal or above 1$ are acceptable since eBay uses an auction system to sell products.

In [93]:
#we remove rows with price 0

autos.drop(autos[autos['price'] == 0].index, inplace = True)

In [94]:
#Checking the price column after the modifications
autos['price'].describe()

count     48566.000000
mean       5909.404666
std       10120.652480
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      999990.000000
Name: price, dtype: float64

# Exploring the date columns

We have five-date columns date_crawled,ad_created, and last_seen the data type of them is a string. To understand them quantitatively we need to convert them to numerical representation, by using value_counts(normalize=True) fun to understand the distribution of the values.
And the last two columns are registration_year and registration_month are integers so we need to do nothing before exploring them quantitatively.

let's understand the format of the three first date columns

In [95]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [96]:
#include date part only 
#normalize=True>that means instead of returning counts,it will be the 
#percent of each unique value makes up of the whole series

#find the percent of each unique date 
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False)

2016-04-03    0.038607
2016-03-20    0.037887
2016-03-21    0.037372
2016-03-12    0.036919
2016-03-14    0.036548
2016-04-04    0.036486
2016-03-07    0.036013
2016-04-02    0.035477
2016-03-28    0.034860
2016-03-19    0.034777
2016-03-15    0.034283
2016-03-29    0.034119
2016-04-01    0.033686
2016-03-30    0.033686
2016-03-08    0.033295
2016-03-09    0.033089
2016-03-22    0.032986
2016-03-11    0.032574
2016-03-23    0.032224
2016-03-26    0.032204
2016-03-10    0.032183
2016-03-31    0.031833
2016-03-17    0.031627
2016-03-25    0.031606
2016-03-27    0.031092
2016-03-16    0.029609
2016-03-24    0.029342
2016-03-05    0.025326
2016-03-13    0.015669
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012910
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [97]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False)

2016-04-03    0.038854
2016-03-20    0.037948
2016-03-21    0.037578
2016-04-04    0.036857
2016-03-12    0.036754
2016-03-14    0.035189
2016-04-02    0.035148
2016-03-28    0.034983
2016-03-07    0.034736
2016-03-29    0.034057
2016-03-15    0.034016
2016-04-01    0.033686
2016-03-19    0.033686
2016-03-30    0.033501
2016-03-08    0.033315
2016-03-09    0.033151
2016-03-11    0.032904
2016-03-22    0.032801
2016-03-26    0.032265
2016-03-23    0.032059
2016-03-10    0.031895
2016-03-31    0.031874
2016-03-25    0.031751
2016-03-17    0.031277
2016-03-27    0.030989
2016-03-16    0.030124
2016-03-24    0.029280
2016-03-05    0.022897
2016-03-13    0.017008
2016-03-06    0.015319
                ...   
2016-02-18    0.000041
2016-02-20    0.000041
2016-02-02    0.000041
2016-02-14    0.000041
2016-01-10    0.000041
2016-02-24    0.000041
2016-02-05    0.000041
2016-02-26    0.000041
2016-02-12    0.000041
2015-09-09    0.000021
2015-11-10    0.000021
2016-01-03    0.000021
2016-01-16 

The highest number of ads created was in two months (March and April).

In [98]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False)

2016-04-06    0.221801
2016-04-07    0.131944
2016-04-05    0.124758
2016-03-17    0.028085
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024770
2016-04-04    0.024482
2016-03-12    0.023782
2016-03-31    0.023782
2016-04-01    0.022794
2016-03-29    0.022361
2016-03-22    0.021373
2016-03-28    0.020858
2016-03-20    0.020652
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018531
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015875
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012601
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

About (22%) of the ads had been seen online was in April/2016

let's now explore the last two date columns registration_year and registration_month

In [99]:
autos['registration_year'].describe()

count    48566.000000
mean      2004.755014
std         88.643020
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The max date is 9999 is in the future!
The min date is 1000, before the cars were invented!
Based on Google the first automobile was in 1885–1886. So any registration year before that date is unacceptable, and any year after 2016 is an inaccurate date since the date of crawling is in 2016.

We will count the number of listings with cars that fall outside the 1885 - 2016 interval to see if it is safe to remove them

In [105]:
outlier_year_less = autos[autos['registration_year'] < 1885]

outlier_year_more = autos[autos['registration_year'] > 2016]

In [106]:
outlier_year_more.shape[0]


1879

In [107]:
outlier_year_less.shape[0]

5

The total outliers = 1879 + 5 = 1884 (which is ~3% of the total data set) 
The total data set being 50,000 values. 

Hence we will remove it. 

In [109]:
autos.drop(outlier_year_more.index, inplace = True)

In [110]:
autos.drop(outlier_year_less.index, inplace = True)

Now lets check the registration_month

In [111]:
autos['registration_month'].describe()

count    46682.000000
mean         5.827257
std          3.670371
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [117]:
autos['registration_month'].sort_index(ascending= False).value_counts()

3     4838
6     4119
0     4026
4     3894
5     3877
7     3722
10    3495
12    3263
9     3243
11    3229
1     3105
8     3021
2     2850
Name: registration_month, dtype: int64

We see a month value of 0, which is the 3rd highest in number

We cant just remove it as we'll lose useful data, hence we'll replace the 0 values with null

In [118]:
#replacing 0 with null
autos.loc[autos['registration_month'] == 0,'registration_month'] = 'null'

In [120]:
autos['registration_month'].value_counts(normalize = True)

3       0.103637
6       0.088235
null    0.086243
4       0.083415
5       0.083051
7       0.079731
10      0.074868
12      0.069898
9       0.069470
11      0.069170
1       0.066514
8       0.064714
2       0.061051
Name: registration_month, dtype: float64

10% of the cares were registered in the third month (March)

In [125]:
brands_ = autos['brand'].value_counts(normalize=True)
brands_

volkswagen        0.211281
bmw               0.110042
opel              0.107579
mercedes_benz     0.096461
audi              0.086564
ford              0.069898
renault           0.047149
peugeot           0.029840
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008761
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

Most brands don't have a significant percentage in the listing...so we will keep brands that have more than 5% of total listings.

In [127]:
brands_more_than_5 = brands_[brands_> 0.05].index
brands_more_than_5

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

**Lets calculate the average price for each of the top brands**

In [132]:
avg_price = {}

for b in brands_more_than_5:
    mean_ = autos.loc[autos['brand'] == b, 'price'].mean()
    avg_price[b] = int(mean_)
    
avg_price

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5503}

There is a significant gap in average price between these common and top brands. volkswagen's price has the middle average..while audi,mercedes_benz, and bmw have the highest average.

In [134]:
#get the ave. price for each unique mileage
autos.groupby('odometer_km')['price'].mean()

odometer_km
5000       8873.515924
10000     20550.867220
20000     18448.477089
30000     16608.836842
40000     15499.568381
50000     13812.173212
60000     12385.004433
70000     10927.182814
80000      9721.947636
90000      8465.025105
100000     8132.697279
125000     6214.022030
150000     3801.039587
Name: price, dtype: float64

We see an expected pattern (inverse relationship), when the mileage is increasing the ave. price is decreasing.

# Cars With Damage VS Non-damaged (Price)

How much cheaper are cars with damage than their non-damaged ?

In [140]:
autos.groupby('unrepaired_damage')['price'].mean()


unrepaired_damage
ja      2460.866109
nein    7164.033103
Name: price, dtype: float64

In [142]:
#Difference between repared and unrepaired cars in price
autos.groupby('unrepaired_damage')['price'].mean()[1] - autos.groupby('unrepaired_damage')['price'].mean()[0]

4703.1669940093925

Damaged cars are cheaper than non-damaged about (5000$)

# Conclusion:

- Between the most common brands, there is a significant gap in the average prices. The most expensive brands are Audi, Mercedes-Benz, and BMW.
- Ford and Opel have the lowest average price among the top brands, while Volkswagen gets a moderate average price.
- There is a pattern between mileage and average price. When the mileage is increasing the average price is decreasing (inverse relationship).
- Cars with damage are cheaper than non-damaged about (5000$)


Key Takeaways:
- grouby method in panda
- cleaning column names
- removing outlier values
